### PR TITLE
[Crypto] Add SubtleCrypto APIs for HTTP and gRPC

### PR DESCRIPTION
**Pull request by ItalyPaleAle**
> Spin-off from #5429

Implements the Subtle Crypto APIs (low-level) as defined in https://github.com/dapr/proposals/pull/3

These APIs are implemented as universal APIs. They include support for resiliency and metrics.

### CODE REVIEW

**Review by mukundansundar — COMMENTED** [review]
as discussed yesterday, please remove the public APIs alone for Subtle Crypto

**Comment by ItalyPaleAle** [comment]
PR is now ready for review.

The subtlecrypto APIs are gated behind a build-time flag and are available only if you build Dapr with `-tags subtlecrypto`

Otherwise, all calls to those APIs will return an error indicating they are un-implemented.

File: dapr/proto/runtime/v1/dapr.proto [hunk 1]
```diff
@@ -687,6 +708,174 @@ message UnlockResponse {
...
+message SubtleGetKeyAlpha1Request {
+  enum KeyFormat {
+    // PEM (PKIX) (default)
+    PEM = 0;
+    // JSON (JSON Web Key) as string
+    JSON = 1;
+  }
+ 
+  // Name of the component
+  string component_name = 1 [json_name="component"];
```
Comment (mukundansundar) [inline dapr/proto/runtime/v1/dapr.proto:721]:
```suggestion
  string component_name = 1 [json_name="componentName"];
```
Lets have the same name across both proto and json
same for other scenarios.

File: pkg/grpc/universalapi/api_subtlecrypto_subtlecrypto.go [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+		return &runtimev1pb.SubtleVerifyAlpha1Response{}, err
+	}
+
+	return &runtimev1pb.SubtleVerifyAlpha1Response{
+		Valid: valid,
+	}, nil
+}
+
+// Internal method that checks if the request is for a valid crypto component.
+func (a *UniversalAPI) cryptoValidateRequest(componentName string) (contribCrypto.SubtleCrypto, error) {
```
Comment (mukundansundar) [inline pkg/grpc/universalapi/api_subtlecrypto_subtlecrypto.go:336]:
```suggestion
func (a *UniversalAPI) validateRequest(componentName string) (contribCrypto.SubtleCrypto, error) {
```

File: pkg/grpc/universalapi/api_subtlecrypto_subtlecrypto.go [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+	policyRunner := resiliency.NewRunner[jwk.Key](ctx,
+		a.Resiliency.ComponentOutboundPolicy(in.ComponentName, resiliency.Crypto),
+	)
+	start := time.Now()
+	res, err := policyRunner(func(ctx context.Context) (jwk.Key, error) {
+		return component.GetKey(ctx, in.Name)
+	})
+	elapsed := diag.ElapsedSince(start)
+
+	diag.DefaultComponentMonitoring.CryptoInvoked(ctx, in.ComponentName, diag.Get, err == nil, elapsed)
```
Comment (mukundansundar) [inline pkg/grpc/universalapi/api_subtlecrypto_subtlecrypto.go:55]: diagnostics wise, the key is got here ... but then it can fail to be marshalled leading to an error, shouldn't diagnostics be set only after it is marshalled, just be returning the value?

File: pkg/grpc/universalapi/api_subtlecrypto_subtlecrypto.go [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+			err = fmt.Errorf("failed to marshal public key %s as JSON: %w", in.Name, err)
+			err = messages.ErrCryptoGetKey.WithFormat(in.Name, err)
+			a.Logger.Debug(err)
+			return &runtimev1pb.SubtleGetKeyAlpha1Response{}, err
+		}
+
+	default:
+		err = messages.ErrBadRequest.WithFormat("invalid key format")
+		a.Logger.Debug(err)
+		return &runtimev1pb.SubtleGetKeyAlpha1Response{}, err
```
Comment (mukundansundar) [inline pkg/grpc/universalapi/api_subtlecrypto_subtlecrypto.go:108]: can be moved to before the component is invoked.

File: pkg/grpc/universalapi/api_subtlecrypto_subtlecrypto.go [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+	if err != nil {
+		return &runtimev1pb.SubtleEncryptAlpha1Response{}, err
+	}
+
+	policyRunner := resiliency.NewRunner[subtleEncryptRes](ctx,
+		a.Resiliency.ComponentOutboundPolicy(in.ComponentName, resiliency.Crypto),
+	)
+	start := time.Now()
+	ser, err := policyRunner(func(ctx context.Context) (r subtleEncryptRes, rErr error) {
+		r.ciphertext, r.tag, rErr = component.Encrypt(ctx, in.Plaintext, in.Algorithm, in.Key, in.Nonce, in.AssociatedData)
```
Comment (mukundansundar) [inline pkg/grpc/universalapi/api_subtlecrypto_subtlecrypto.go:134]: can the algorithm be validated before calling the component?

File: dapr/proto/runtime/v1/dapr.proto [hunk 6]
```diff
@@ -687,6 +708,174 @@ message UnlockResponse {
...
+// SubtleEncryptAlpha1Request is the request for SubtleEncryptAlpha1.
+message SubtleEncryptAlpha1Request {
+  // Name of the component
+  string component_name = 1 [json_name="component"];
+  // Message to encrypt.
+  bytes plaintext = 2;
+  // Algorithm to use, as in the JWA standard.
+  string algorithm = 3;
+  // Name (or name/version) of the key.
+  string key = 4;
```
Comment (mukundansundar) [inline dapr/proto/runtime/v1/dapr.proto:746]:
```suggestion
  string key_name = 4 [json_name="keyName"];
```
and similar changes in all other places. I believe we are calling it keyName in other places i.e. components-contrib

File: pkg/messages/predefined.go [hunk 7]
```diff
@@ -95,7 +95,8 @@ const (
 
 var (
 	// Generic.
-	ErrBadRequest = APIError{"invalid request: %v", "ERR_BAD_REQUEST", http.StatusBadRequest, grpcCodes.InvalidArgument}
+	ErrBadRequest     = APIError{"invalid request: %v", "ERR_BAD_REQUEST", http.StatusBadRequest, grpcCodes.InvalidArgument}
+	ErrAPIUnavailable = APIError{"this API is currently not available", "ERR_API_UNAVAILABLE", http.StatusNotImplemented, grpcCodes.Unimplemented}
```
Comment (mukundansundar) [inline pkg/messages/predefined.go:99]: should this more API not implemented?

File: pkg/proto/runtime/v1/dapr_additional.go [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package runtime
+
+// This file contains additional, hand-written methods added to the generated objects.
```
Comment (mukundansundar) [inline pkg/proto/runtime/v1/dapr_additional.go:16]: Why is this file needed?
Is `SetComponentName` not automatically generated?

File: pkg/grpc/universalapi/api_subtlecrypto_subtlecrypto.go (hunk 2 above)
Comment (ItalyPaleAle) [reply pkg/grpc/universalapi/api_subtlecrypto_subtlecrypto.go:336]: I'll leave it to `crypto` because the struct is used by a number of APIs, not just crypto, and other building blocks may have their own `validateRequest`.

File: dapr/proto/runtime/v1/dapr.proto (hunk 6 above)
Comment (ItalyPaleAle) [reply dapr/proto/runtime/v1/dapr.proto:746]: Changed

File: dapr/proto/runtime/v1/dapr.proto (hunk 1 above)
Comment (ItalyPaleAle) [reply dapr/proto/runtime/v1/dapr.proto:721]: Changed

(Although the JSON name here is unused)

File: pkg/grpc/universalapi/api_subtlecrypto_subtlecrypto.go (hunk 3 above)
Comment (ItalyPaleAle) [reply pkg/grpc/universalapi/api_subtlecrypto_subtlecrypto.go:55]: Looks kike this is consistent with what we do for other building blocks: https://github.com/ItalyPaleAle/dapr/blob/e07039b316f2ff8cfc0558d42c73e9054a539789/pkg/grpc/api.go

I think this is fine as what we want to measure here is the success of the operation from the component and the latency it adds

File: pkg/messages/predefined.go (hunk 7 above)
Comment (ItalyPaleAle) [reply pkg/messages/predefined.go:99]: Changed

File: pkg/proto/runtime/v1/dapr_additional.go (hunk 8 above)
Comment (ItalyPaleAle) [reply pkg/proto/runtime/v1/dapr_additional.go:16]: It is not.

I need these methods so I can have an interface that matches those structs, and allows avoiding repetition of a lot of code.

File: pkg/grpc/universalapi/api_subtlecrypto_subtlecrypto.go (hunk 5 above)
Comment (ItalyPaleAle) [reply pkg/grpc/universalapi/api_subtlecrypto_subtlecrypto.go:134]: Are you thinking of using the `SupportedEncryptionAlgorithms` method?

Comment (mukundansundar) [reply pkg/grpc/universalapi/api_subtlecrypto_subtlecrypto.go:134]: Yes

Comment (ItalyPaleAle) [reply pkg/grpc/universalapi/api_subtlecrypto_subtlecrypto.go:134]: IMHO that's redundant. All components right now synchronously check if the algorithm is supported at the very beginning. In the case of AKV, it's checked before invoking the cloud service: https://github.com/dapr/components-contrib/blob/master/crypto/azure/keyvault/component.go#L147

I'd let the component make that decision personally. There can also be edge cases the component can handle better than the runtime.

File: pkg/grpc/universalapi/api_subtlecrypto.go [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+	return nil, messages.ErrAPIUnimplemented
+}
+
+// SubtleVerifyAlpha1 verifies the signature of a message using a key stored in the vault.
+func (a *UniversalAPI) SubtleVerifyAlpha1(ctx context.Context, in *runtimev1pb.SubtleVerifyAlpha1Request) (*runtimev1pb.SubtleVerifyAlpha1Response, error) {
+	return nil, messages.ErrAPIUnimplemented
+}
+
+// CryptoValidateRequest is an internal method that checks if the request is for a valid crypto component.
+func (a *UniversalAPI) CryptoValidateRequest(componentName string) (contribCrypto.SubtleCrypto, error) {
```
Comment (mukundansundar) [inline pkg/grpc/universalapi/api_subtlecrypto.go:63]: nitpick: any reason to export the function here?

Comment (mukundansundar) [reply pkg/grpc/universalapi/api_subtlecrypto.go:63]: it also seems to be repeated in the file `api_subtlecrypto_subtlecrypto.go` 
https://github.com/dapr/dapr/pull/6137/files#diff-d1155b1f22248cfd3a8bafe9904e5b41a702cbe4dca72c91d50f216b8f7caa7cR340

Comment (ItalyPaleAle) [reply pkg/grpc/universalapi/api_subtlecrypto.go:63]: Yes It's needed in #6197 

It's repeated because `api_subtlecrypto_subtlecrypto.go` requires the build tag `subtlecrypto`, while `api_subtlecrypto.go` is built only if the build tag is NOT present.

**Review by mukundansundar — DISMISSED** [review]
lgtm

File: pkg/http/api_subtlecrypto.go [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+	)
+}
+
+func (a *api) onPostSubtleCryptoSign() fasthttp.RequestHandler {
+	return UniversalFastHTTPHandler(
+		a.universal.SubtleSignAlpha1,
+		UniversalFastHTTPHandlerOpts[*runtimev1pb.SubtleSignAlpha1Request, *runtimev1pb.SubtleSignAlpha1Response]{
+			InModifier: subtleCryptoInModifier[*runtimev1pb.SubtleSignAlpha1Request],
+		},
+	)
```
Comment (mukundansundar) [inline pkg/http/api_subtlecrypto.go:120]: @ItalyPaleAle adding unit tests for these apis would be good

Comment (ItalyPaleAle) [reply pkg/http/api_subtlecrypto.go:120]: When they get enabled, yes :) Right now, there's not much to test

**Review by mukundansundar — APPROVED** [review]
